### ISSUE TITLE
Entry Editor only opens after opening Preferences dialog

### ISSUE THREAD

**Issue by matheger** [JabRef/jabref#4762]
JabRef version 5.0-dev on Windows 10

- [x] I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

When loading an existing bibliography, the Entry Preview end Entry Editor panes only work after opening the Preferences and clicking "Save". (I don't even need to change any settings - just opening the dialog is enough.) This happens every time a new bibliography is opened, ie. even without restarting the program in the meantime.

1. Create a new bibliography and add an entry
2. Save and close bibliography
3. Reopen
4. Entry Editor fails to show on double-click on the entry
5. Open Options -> Preferences
6. Hit "Save"
7. Entry Editor now opens on double click

The Event Log only shows "Can only load style files: Preview"

**Comment by AEgit** [comment]
Can confirm this issue for:

JabRef 5.0-dev--snapshot--2019-03-15--master--7ed15bc4a
Windows 10 10.0 amd64 
Java 1.8.0_201

**Comment by JoKalliauer** [comment]
Have the same issue on Ubuntu for:

JabRef 5.0-dev--snapshot--2019-03-15--master--7ed15bc4a
Linux 4.15.0-46-generic amd64 
Java 1.8.0_201

**Comment by 8greg8** [comment]
I found out that if you go to `Preferences>Entry editor` and uncheck `Open editor when a new entry is created` everything works then.

**Comment by AEgit** [comment]
That might be a workaround, but the bug remains

**Comment by matheger** [comment]
> I found out that if you go to `Preferences>Entry editor` and uncheck `Open editor when a new entry is created` everything works then.

Nope, not for me.

**Comment by bernhard-kleine** [comment]
JabRef 5.0-dev--snapshot--2019-03-15--master--46d2c7dc2
Windows 7 6.1 amd64 
Java 1.8.0_201

Not only the entry editor is gone, but also the entry preview. Latest version of today. Is Jabref becoming Bananaware, matures at the client?

**Comment by Siedlerchr** [comment]
Does resetting the preferences help?

**Comment by matheger** [comment]
> Does resetting the preferences help?

No, this also happens after a fresh install.

**Comment by bernhard-kleine** [comment]
@Siedlerchr: It does

**Comment by AEgit** [comment]
@Siedlerchr: Resetting the preferences does not help (at least for me).

JabRef 5.0-dev--snapshot--2019-03-15--master--7ed15bc4a
Windows 10 10.0 amd64 
Java 1.8.0_201

**Comment by Siedlerchr** [comment]
I can confirm this error also on Windows.
  If you open the prefences dialog and click on save again, entry editor and preview will work again.
I am currently trying to figure out why

**Comment by AEgit** [comment]
Note, that even before this bug the preview would not work/turn up before having started the entry editor at least once (I think there was on issue on this before, which is still open). The difference now is just, that you cannot even open the entry editor with, e. g., doubleclick, before having used the "Preferences" workaround.

JabRef 5.0-dev--snapshot--2019-03-15--master--7ed15bc4a
Windows 10 10.0 amd64 
Java 1.8.0_201

**Comment by Siedlerchr** [comment]
@AEgit  okay, thanks for your feedback. This might narrow it down to the preview initializing

**Comment by Siedlerchr** [comment]
@tobiasdiez I first thought it has something to do with the preview citation style. I debugged and  temporarily added the discovery of the citation styles with a future.get() in the preview panel, but does not change anything.
And the workaround only works if you hit save in the preferences dialog. 
So there must be a thing which prevents the entry editor from opening which is set in the preferences when you click save.  I absolutely have no idea.
A resetting of preferences does not change anything

**Comment by matthiasgeiger** [comment]
Maybe 'git bisect' helps to track it down?

**Comment by tobiasdiez** [comment]
That's strange. The DatabaseChangePane apparently hides the entry preview/editor. I don't have time the next few days to investigate this further, but will return to it on Monday the latest.

**Comment by wujastyk** [comment]
JabRef 5.0-dev--snapshot--2019-03-16--master--d7dd41b00
Linux 4.15.0-46-generic amd64 
Java 1.8.0_201

Same problem as reported above.  Options / Prefernces / Save  re-enables editing and preview.

**Comment by Siedlerchr** [comment]
I further found out that it only works if you click OK/Save in the Preferences dialog. Simply opening the dialog and pressing Cancel or Escape does not work.

**Comment by CaptainDaVinci** [comment]
Is this issue result of the changes made [recently](https://github.com/JabRef/jabref/pull/4693/files#diff-78d6fdd24f8cf9ec9fe6d26f79b6f22e) to the ` src/main/java/org/jabref/gui/BasePanel.java` file? Undoing few of those changes in the constructor seem to fix the issue.

**Comment by Siedlerchr** [comment]
Hmm that's interesting, thanks for your investigation. The reason why it fails would be interesting

**Comment by CaptainDaVinci** [comment]
The reason why I think it works on clicking _Save_ under preferences dialog would be because that results in setting up the main panel once again which in turn clears the `getChildren()` for `BasePanel`. However, on creating an instance of `BasePanel`, the constructor makes a call to setting up the main panel before adding `changePanel` to it.

**Comment by Siedlerchr** [comment]
I think I understand the problem. It's a Stack pane. So there can be only at most one child visible. https://o7planning.org/en/10633/javafx-stackpane-layout-tutorial

**Comment by bernhard-kleine** [comment]
(@Siedlerchr) In addition to the errors reported earlier, the Ctrl-E does not work even after Open preferences->Save. You have to start the editor by View->Open entry editor.

**Comment by CaptainDaVinci** [comment]
> In addition to the errors reported earlier, the Ctrl-E does not work even after Open preferences->Save. You have to start the editor by View->Open entry editor.

 This issue remains even after removing the line which adds the `changePane` to the `StackPane`. In addition to that, keybinding for _toggle preview_  isn't working.

However, once the you display the editor from view -> Open Entry Editor the keybindings work.

**Comment by zoltanpm** [comment]
> @tobiasdiez I first thought it has something to do with the preview citation style. I debugged and temporarily added the discovery of the citation styles with a future.get() in the preview panel, but does not change anything.

I agree -- I get the following in the log:

```
Can only load style files: Preview
Can only load style files: Preview
```

After open/save preferences, it works fine; 5.0-dev on Fedora 29.

**Comment by tobiasdiez** [comment]
The main bug should be fixed in the latest development version. Could you please check the build from http://builds.jabref.org/master/. Thanks! Please remember to make a backup of your library before trying-out this version.

There are still some related bugs left that should be fixed as well: https://github.com/JabRef/jabref/pull/4784#pullrequestreview-216419250

**Comment by wujastyk** [comment]
JabRef 5.0-dev--snapshot--2019-03-19--master--90376f40f
Linux 4.18.0-16-generic amd64 
Java 1.8.0_201

Fixed for me.  Double click n a row now opens the editor.

**Comment by bernhard-kleine** [comment]
the Alt-2 key to show the entry preview still is not working on a freshly loaded library. the Ctrl-E, however, works, and after opening the entry editor, Alt-2 works, too. Very strange.

**Comment by CaptainDaVinci** [comment]
<kbd>ctrl</kbd>+<kbd>e</kbd> doesn't seem to work for me. Even after opening the editor <kbd>ctrl</kbd>+<kbd>e</kbd> only closes the editor and does not toggle.

**Comment by bernhard-kleine** [comment]
@CaptainDaVinci : hier JabRef 5.0-dev--snapshot--2019-03-19--master--90376f40f on Windows 7 6.1 amd64 with Java 1.8.0_201
toggle works

**Comment by AEgit** [comment]
I can confirm that the main bug has been fixed in:

JabRef 5.0-dev--snapshot--2019-03-19--master--90376f40f
Windows 10 10.0 amd64 
Java 1.8.0_201

i. e., it is now possible again to open the entry editor by double-clicking on a row in the main table.
<kbd>CTRL</kbd> + <kbd>E</kbd> does not seem to (reliably) work for me (I only managed to open the entry editor with <kbd>CTRL</kbd> + <kbd>E</kbd> once and so far I haven't managed to reproduce this behaviour): Generally it does not seem possible to open the entry editor using <kbd>CTRL</kbd> + <kbd>E</kbd>. However, I am a bit unsure about this, since I haven't used this keyboard shortcut at all (but I checked in "Customize key bindings" that it was active) - it might be, that something else is off and/or there is a more general problem with the keybindings (?). So this needs more people to confirm.

What does remain, however, - as reported already by @bernhard-kleine - is that the entry preview will not show on a freshly opened database. Only after having opened the entry editor at least once, the entry preview will appear. Should a separate issue be created for this? I thought this problem had already been reported, but I cannot find the issue anymore.

**Comment by Siedlerchr** [comment]
Thanks to  @matthiasgeiger  this issue is now finally resolved.

**Comment by AEgit** [comment]
Cannot confirm for:

JabRef 5.0-dev--snapshot--2019-03-25--master--e14b56742
Windows 10 10.0 amd64 
Java 1.8.0_201

As mentioned in https://github.com/JabRef/jabref/issues/4762#issuecomment-474752885, the main bug reported herein has been fixed. But the other issue reported in https://github.com/JabRef/jabref/issues/4762#issuecomment-474752885 remain - are these considered to be separate issues that should get their own thread? Or are they considered to be fixed? If the latter is the case, I cannot confirm the bugfix for JabRef 5.0-dev--snapshot--2019-03-25--master--e14b56742

**Comment by Siedlerchr** [comment]
@AEgit  The PR https://github.com/JabRef/jabref/pull/4806 fixed the following remaining issues:

> 
>     table is now working again in case a new file is opened (also fixes #4793)
>     in case preferences are saved entry editor is working properly again
>     external changes are now also detected in case file name has changed due to "Save As"

View -> Entry Preview is checked: Uncheck and check again: Preview is shown, 
Keybinding: <kbd>CTRL</kbd>+ <kbd>e</kbd> worked fine for me until I reset the keybindings. This is odd now. I will create a new issue

**Comment by matthiasgeiger** [comment]
You are right, that there is still something wrong with the opening of the preview and entry editor - but this was already wrong before the topic of this issue has been introduced (e.g. in JabRef 5.0-dev--snapshot--2018-12-14--master--49c3ef396 which I have just tested).

The behavior seems to be: The preview is only opened after "toggle preview" twice. The Keybinding seems only to work when the preview is visible...

I'll create another issue for this...

**Comment by AEgit** [comment]
Cheers!
